src/language-server/language-server.ts

```diff
@@ -174,6 +184,62 @@ export default class LanguageServer extends AutoDisposable {
       this._settings = Object.assign({}, defaultSettings, settings);
       this._settingsChanged(this._settings, previousSettings);
     });
+
+    this._connection.onCodeAction(async(req) => {
+      return this.handleErrors(this.getCodeActions(req), []);
+    });
+
+    this._connection.onExecuteCommand(async(req) => {
+      if (req.command === applyEditCommandName) {
+        return this.handleErrors(
+            this.executeApplyEditCommand(req.arguments as [WorkspaceEdit]),
+            undefined);
+      }
+    });
+  }
+
+  private async getCodeActions(req: CodeActionParams) {
+    const commands: Command[] = [];
+    if (req.context.diagnostics === []) {
+      // Currently we only support code actions on Warnings,
+      // so we can early-exit in the case where there aren't any.
+      return commands;
+    }
+    const warnings = await this._editorService.getWarningsForFile(
+        this.converter.getWorkspacePathToFile(req.textDocument));
+    const range = this.converter.convertLRangeToP(req.range, req.textDocument);
+    for (const warning of warnings) {
+      if ((!warning.fix && !warning.actions) ||
+          !isRangeInside(warning.sourceRange, range)) {
```

Comment:
curiosity: how could this happen?

Reply:
ah! Yeah the question was regarding the `isRangeInside` check, and wasn't clear that it's done to filter out warnings from other parts of the file. Would be great to have a comment explaining this 🙏